pkg/io/v2/providers/providers.go

```diff
@@ -0,0 +1,235 @@
+	case providerGS:
+		return getGCSBucket(ctx, credentials, bucket)
+	case providerS3:
+		return getS3Bucket(ctx, credentials, bucket)
+	default:
+		return nil, fmt.Errorf("unknown storageProvider: %s", storageProvider)
+	}
+}
+
+func getFileBucket(bucket string) (*blob.Bucket, error) {
+	bkt, err := blob.OpenBucket(context.Background(), "file://"+bucket)
+	if err != nil {
+		return nil, fmt.Errorf("error opening file bucket: %v", err)
+	}
+	return bkt, nil
+}
+
+func getGCSBucket(ctx context.Context, credentials []byte, bucketName string) (*blob.Bucket, error) {
+	googleCredentials, err := google.CredentialsFromJSON(ctx, credentials, storage.ScopeFullControl)
+	if err != nil {
+		return nil, fmt.Errorf("error getting Google credentials from JSON: %v", err)
+	}
+
+	client, err := gcp.NewHTTPClient(
+		gcp.DefaultTransport(),
+		gcp.TokenSource(googleCredentials.TokenSource))
+	if err != nil {
+		return nil, fmt.Errorf("error creating GCP Http Client: %v", err)
+	}
+
+	bkt, err := gcsblob.OpenBucket(ctx, client, bucketName, nil)
+	if err != nil {
+		return nil, fmt.Errorf("error opening GCS bucket: %v", err)
+	}
+	return bkt, nil
+}
+
+type s3Credentials struct {
+	Region           string `json:"region"`
+	Bucket           string `json:"bucket"`
+	Endpoint         string `json:"endpoint"`
+	Insecure         bool   `json:"insecure"`
+	S3ForcePathStyle bool   `json:"s3_force_path_style"`
+	AccessKey        string `json:"access_key"`
+	SecretKey        string `json:"secret_key"`
+}
+
+func getS3Bucket(ctx context.Context, creds []byte, bucketName string) (*blob.Bucket, error) {
+	s3Credentials := &s3Credentials{}
+	if err := json.Unmarshal(creds, s3Credentials); err != nil {
+		return nil, fmt.Errorf("error getting S3 credentials from JSON: %v", err)
+	}
+
+	staticCredentials := credentials.NewStaticCredentials(s3Credentials.AccessKey, s3Credentials.SecretKey, "")
+
+	sess, err := session.NewSession(&aws.Config{
+		Credentials:      staticCredentials,
+		Endpoint:         aws.String(s3Credentials.Endpoint),
+		DisableSSL:       aws.Bool(s3Credentials.Insecure),
+		S3ForcePathStyle: aws.Bool(s3Credentials.S3ForcePathStyle),
+		Region:           aws.String(s3Credentials.Region),
+	})
+	if err != nil {
+		return nil, fmt.Errorf("error creating S3 Session: %v", err)
+	}
+
+	bkt, err := s3blob.OpenBucket(ctx, sess, bucketName, nil)
+	if err != nil {
+		return nil, fmt.Errorf("error opening S3 bucket: %v", err)
+	}
+	return bkt, nil
+}
+
+func SignedURL(ctx context.Context, credentials []byte, storagePath string, opts *blob.SignedURLOptions) (string, error) {
+	storageProvider, bucketName, relativePath, err := ParseStoragePath(storagePath)
+	if err != nil {
+		return "", err
+	}
+
+	switch storageProvider {
+	case providerFile, providerS3:
+		bucket, err := GetBucket(ctx, credentials, storagePath)
+		if err != nil {
+			return "", err
+		}
+		return bucket.SignedURL(ctx, relativePath, opts)
+	case providerGS:
+		if len(credentials) == 0 {
```

Comment:
Why do we implement handling for unauthenticated usage here rather than in `GetBucket`?

Reply:
Looks like gocloud (and I think also the Google GCS lib) is not able to do this:
```go
	newClient := &gcp.HTTPClient{
		Client: http.Client{
			Transport: &oauth2.Transport{
				Base: gcp.DefaultTransport(),
			},
		},
	}
	bkt, _ := gcsblob.OpenBucket(context.Background(), newClient, "gs://k8s-conformance-openstack", nil)
	test, err := bkt.SignedURL(context.Background(),"README.md", nil)
	fmt.Println(err)
```
Leads to this error:
```
blob (key "README.md") (code=Unknown): to use SignedURL, you must call OpenBucket with a valid Options.GoogleAccessID and exactly one of Options.PrivateKey or Options.SignBytes
```
Which is produced here: https://github.com/google/go-cloud/blob/master/blob/gcsblob/gcsblob.go#L598-L600

Looking at the documentation I think signed URLs are not really meanted to be used on public files with anonymous access: https://cloud.google.com/storage/docs/access-control/signing-urls-manually

Reply:
Grepping to the code I see a single usage of `SignedURL` in prow/spyglass/gcsartifact_fetcher.goL77 and that very explicitly uses auth, so why is this needed?

Reply:
I integrated the stuff from here:
https://github.com/kubernetes/test-infra/blob/master/prow/spyglass/gcsartifact_fetcher.go#L150-L157

But we can discuss/remove it on the next PR if you want